scheduler/src/cook/postgres.clj

```diff
@@ -0,0 +1,98 @@
+           {:user username})))
+
+
+; With unit tests, we need to create a new configuration dictionary in the test fixture.
+; That gets stored here.
+(def saved-pg-config-dictionary (atom nil))
+
+
+(let [custom-formatter (clj-time.format/formatter "yyyyMMdd")]
+(defn make-new-schema-name
+  "Generate the schema name we want to use for unit tests"
+  []
+  (str "db_tests_" (System/getenv "USER") "_" (clj-time.format/unparse-local-date custom-formatter (clj-time.core/today)) "_" (+ 1000 (.nextInt (Random.) 9000)))))
+
+(defn configure-database-connection-for-unit-tests
+  "Configure the database connection."
+  []
+  (if-let [schema-name (System/getenv "COOK_DB_TEST_PG_SCHEMA")]
+    (do
+      (let [pg-connection-meta (make-database-connection-dictionary-for-unit-tests schema-name)]
+        (log/info "Using existing schema " schema-name " with connection " pg-connection-meta)
+        (reset! saved-pg-config-dictionary pg-connection-meta)
+        pg-connection-meta))
+    ; If I have an environment saying to autocreate schema:
+    (do
+      (let [setup-script (System/getenv "COOK_DB_TEST_AUTOCREATE_SCHEMA")
+            schema-name (make-new-schema-name)
+            pg-connection-meta (make-database-connection-dictionary-for-unit-tests schema-name)]
+        (log/info "Auto-creating schema " schema-name "with connection" pg-connection-meta)
+        (setup-database setup-script schema-name)
+        (reset! saved-pg-config-dictionary pg-connection-meta)
+        pg-connection-meta))))
+
+(defn deconfigure-database-connection-for-unit-tests
+  "Deconfigure the database connection. Used when completing a test fixture."
+  []
+  (when-not (System/getenv "COOK_DB_TEST_PG_SCHEMA")
+    (let [schema-name (:currentSchema @saved-pg-config-dictionary)]
+      (log/info "Auto-destroying schema " schema-name "with connection" @saved-pg-config-dictionary)
+      ; Can't use '?' here with parameter substitution.
+      (sql/execute! @saved-pg-config-dictionary [(str "DROP SCHEMA " schema-name " CASCADE;")]))
+    (reset! saved-pg-config-dictionary nil)))
+
+(defn with-pg-db [f]
+  "Test fixture that sets up a new postgres database (if needed), runs the unit tests, and demolishes it afterwards. Setup a postgres database test fixture that creates a database and destroys it afterwards"
+  (configure-database-connection-for-unit-tests)
+  (try
+    (f)
+    (finally
+      (deconfigure-database-connection-for-unit-tests))))
+
+(defn pg-db
+  "Return access information for a postgresql database suitable for database access.
+
+   There are 2 main codepaths we take here:
+
+   1. If there is a config/config with a pg-config in config.edn, (I.e., we're running
+      integration tests or production) run with that config.
+
+   2. If there's a saved configuration dictionary (used for unit tests), return that."
+  []
+  (or
+    (-> config/config :settings :pg-config)
+    @saved-pg-config-dictionary))
```

Comment:
Same question as above for most things in this file, I suppose: should this stuff be under `scheduler/test/` instead of `scheduler/src/`?

Reply:
Nope. This is how all of the code in Cook gets a database connection. See E.g., https://github.com/scrosby/Cook/blob/wip/postgres-docker/scheduler/src/cook/quotashare.clj#L146

Reply:
Most of the things in this file seem unit-test-specific.

Reply:
Yes. Because we're having Cook create DDL in test fixtures, most of the complexity in this file is related to unit tests. Had we went with the 'database is assumed to exist' this file would have `make-database-connection-dictionary-for-unit-tests` + 4 lines of code because unit tests wouldn't have needed all of these extra codepaths.  In addition, this is going to get more coupled, stateful, and need lifecycle management when we add connection pooling.